bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/graphics/GC.java

```diff
@@ -1807,14 +1807,25 @@ private void doFastDrawText(String string, int x, int y) {
 }
 
 private GCTextData getTextData(String string) {
-	GCTextData data = (GCTextData) textDataCache.get(string);
-	if (data == null) {
+	// Create a cache key from the string, alpha, font and foreground color
+	// This ensures that the same string with different attributes is drawn correctly
+	StringBuilder str = new StringBuilder(string).append(data.alpha);
+	if(data.font != null) {
+		str.append(data.font.hashCode());
```

Comment:
`hashCode()` is not collission-free and therefore should not be used as a unique key.

Instead, I propose to either use a nested `Map<String, Map<Font>, Map<Color>>>` or create a small wrapper `Record` combining the 3 attributes and use that as key.

Reply:
Not sure how to do the nested Map. For the wrapper `Record` do you mean something like this?

```java
record Key(String string, int alpha, Font font, double[] fgColor) {}

String key = new Key(string, data.alpha, data.font, data.foreground).toString();
```

I tried it and it's a bit slower than using the `hashCode()` method. I used `hashCode()` to make the key String a bit shorter.

Reply:
A I only used `hashCode()` to make the key String a bit shorter could we just not use the objects? 

```java
StringBuilder str = new StringBuilder(string).append(data.alpha);
if(data.font != null) {
	str.append(data.font);
}
if(data.foreground != null) {
	str.append(data.foreground);
}
String key = str.toString();
```

Reply:
> Not sure how to do the nested Map. 

```java
Map<String, Map<Font, Map<Color, GCData>>> cache = new HashMap<>();
GCData gcdata = cache.computeIfAbsent(string, k -> new HashMap<>()).computeIfAbsent(font, k -> new HashMap<>()).get(color);
```

> For the wrapper Record do you mean something like this?
> ```java
> record Key(String string, int alpha, Font font, double[] fgColor) {}
> 
> String key = new Key(string, data.alpha, data.font, data.foreground).toString();
> ```

Yes, but without `toString()`.

```java
record Key(String string, int alpha, Font font, double[] fgColor) {}
Map<Key, GCData> cache;
Key key = new Key(string, data.alpha, data.font, data.foreground);
cache.get(key);
```

> I tried it and it's a bit slower than using the hashCode() method.

Do you have numbers how much slower?

> I used hashCode() to make the key String a bit shorter.

`hashCode()` is semantically incorrect. Two `Font`s or `Color`s hash codes may collide and be incorrectly mapped to the same cache entry. It's quite unlikely, but it's very bad practice to use `hashCode()` for something like this and we should avoid it.

Reply:
> A I only used `hashCode()` to make the key String a bit shorter could we just not use the objects?

In practice I suppose we could, but that would rely on on its `toString()` implementation to return something unique and it's probably equally expensive.


Another problem I am not sure of, with the solution of holding on to the concrete references in `Key` is that we effectively preventing garbage collection of those `Font`s and `Color`s.

Using `WeakHashMap`s should solve that:

```java
Map<String, Map<Font, Map<Color, GCData>>> cache = new WeakHashMap<>();
GCData gcdata = cache.computeIfAbsent(string, k -> new WeakHashMap<>()).computeIfAbsent(font, k -> new WeakHashMap<>()).get(color);
```

Reply:
> Yes, but without toString().

I force pushed this now and it's as fast as my previous tests (using [test snippet](url)).

> Another problem I am not sure of, with the solution of holding on to the concrete references in Key is that we effectively preventing garbage collection of those Fonts and Colors.

When the GC is destroyed the Cache is cleared with `textDataCache.release()` - is that sufficient?

Reply:
> Actually, it's only the `Font` that is an object as we are storing an array of doubles for the foreground color.

For the `Font`:  We could also just hold on to the primitive (`long`) `font.handle.id` instead of the whole `Font` reference.

For the `Color`: `record#equals(Object)` does compare array members using `==`, but I guess this is OK here, it's probably not happening that we are dealing with two equal - but not same - `Color` references.

Reply:
> For the Font: We could also just hold on to the primitive (long) font.handle.id instead of the whole Font reference.

I'm not sure if `font.handle` could be null. Here's `Font#hashCode()`

```java
public int hashCode() {
	return handle != null ? (int)handle.id : 0;
}
```

Reply:
I've force pushed a new commit that uses `data.font.handle.id` for the font with null checks for `data.font` and `data.font.handle`. The key creation is a bit more verbose but now the cache key only contains primitives, not references to objects.

Reply:
I've just noticed in the build failure reports:

```
warning: 'record' may become a restricted type name in a future release and may be unusable for type declarations or as the element type of an array
		private static record Key(String string, int alpha, long font, double[] fgColor) {}
```

Reply:
> Is something in the build process not Java 17 compliant?

Well, ant still executed on Java 11 (see https://github.com/eclipse-platform/eclipse.platform.swt/issues/626) but the compiler should be able to compile Java 17 constructs if properly configured.

So far it looks like it is using plain `javac` task with almost no configuration for target/source/release ???
See https://github.com/eclipse-platform/eclipse.platform.swt/blob/2b9b6b4773991b9b15d534bffd660d1777a76238/bundles/org.eclipse.swt/buildSWT.xml#L298

If that is correct (and no other magic property is used), ant uses Java 11 and Java 11 uses Java 11 as default target, so it is correct to not being able to use records.

@HannesWell : how far are your efforts to get rid of ant in SWT?

Reply:
> so it is correct to not being able to use records.

OK, thanks for that info.

So it seems that, at least for now, we cannot use a Java `record` for the cache key. A Java class could be used instead but then we'd have to supply our own `hashCode()` implementation.

Reply:
> @HannesWell : how far are your efforts to get rid of ant in SWT?

Getting entirly rid of ANT in SWT will probably take some more time.
However I had a quick look and got the impression that at the moment the SWT java sources for all platforms are build twice with ant to verify that there are no compilation failures. But since the java code is already compiled with maven that seems useless for me now. So I would solve this particular problem by just removing the two extra ANT compilations.

I won't have time to check and work that out in detail before Sunday, probably even Monday, but I will do it with top priority then.

Reply:
PR https://github.com/eclipse-platform/eclipse.platform.swt/pull/661 is merged.
You now should be able to use Java-17 language features. In the mentioned PR I verified that it works.

Reply:
@HannesWell Thanks for that, this PR now builds and compiles.

@sratz What do you think? The cache key now consists only of the primitives of the string, alpha, font handle id and foreground color double[].